File: trie/trie.js
class Trie {
  constructor(char) {
    this.val = char;
    this.children = {};
    this.end = false;
  }

  insert(string) {
    let currentNode = this;
    const { length } = string;
    for (let i = 0; i < length; i += 1) {
      const char = string[i];
      const { children } = currentNode;
      if (!children[char]) children[char] = new Trie(char);
      currentNode = children[char];
    }
    currentNode.end = true;
  }

  search(string) {
    let currentNode = this;
    const { length } = string;
    for (let i = 0; i < length; i += 1) {
      const char = string[i];
      const { children } = currentNode;
      if (!children[char]) return false;
      currentNode = children[char];
    }
    return currentNode.end;
  }

  startsWith(string) {
    let currentNode = this;
    const { length } = string;
    for (let i = 0; i < length; i += 1) {
      const char = string[i];
      const { children } = currentNode;
      if (!children[char]) return false;
      currentNode = children[char];
    }
    return true;
  }
}

module.exports = { Trie };

File: trie/trie.py
class Trie:
    def __init__(self, char: str = ''):
        self.char = ''
        self.children = {}
        self.end = False


    def insert(self, string):
        current = self
        for char in string:
            children = current.children
            if not children.get(char): children[char] = Trie(char)
            current = children[char]
        current.end = True
        return self
    

    def search(self, string):
        current = self
        for char in string:
            children = current.children
            if not children.get(char): return False
            current = children[char]
        return current.end


    def starts_with(self, string):
        current = self
        for char in string:
            children = current.children
            if not children.get(char): return None
            current = children[char]
        return current